lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,549 @@
+
+        auto *defOp = val.getDefiningOp();
+        assert(defOp && "cannot be a BlockArgument here");
+        if (!visited.contains(defOp)) {
+          worklist.push_back(defOp);
+          addedSomething = true;
+        }
+      }
+
+      if (addedSomething)
+        continue;
+
+      auto addToWorklist = [&](Operation *op) {
+        if (op->use_empty())
+          worklist.push_front(op);
+      };
+      auto result = dispatchOpVisitor(curr, addToWorklist);
+      if (failed(result))
+        return failure();
+
+      if (*result == DeletionKind::Delete)
+        toDelete.insert(curr);
+
+      visited.insert(curr);
+      worklist.pop_back();
+    }
+
+    // FIXME: this assumes that we didn't query the opaque value from an
+    // interpreted elaborator value in a way that it can remain used in the IR.
+    for (auto *op : toDelete) {
+      op->dropAllUses();
+      op->erase();
+    }
+
+    // Reduce max memory consumption and make sure the values cannot be accessed
+    // anymore because we deleted the ops above.
+    state.clear();
+    interned.clear();
+
+    return success();
+  }
+
+private:
+  const ElaborationOptions &options;
+  std::mt19937 rng;
+
+  // A map used to intern elaborator values. We do this such that we can
+  // compare pointers when, e.g., computing set differences, uniquing the
+  // elements in a set, etc. Otherwise, we'd need to do a deep value comparison
+  // in those situations.
+  // Use a pointer as the key with custom MapInfo because of object slicing when
+  // inserting an object of a derived class of ElaboratorValue.
+  // The custom MapInfo makes sure that we do a value comparison instead of
+  // comparing the pointers.
+  DenseMap<ElaboratorValue *, std::unique_ptr<ElaboratorValue>, InternMapInfo>
+      interned;
+
+  // A map from SSA values to a pointer of an interned elaborator value.
+  DenseMap<Value, ElaboratorValue *> state;
+
+  SymbolTable symTable;
+};
+
+//===----------------------------------------------------------------------===//
+// Elaborator Pass
+//===----------------------------------------------------------------------===//
+
+class ElaborationPass : public rtg::impl::ElaborationPassBase<ElaborationPass> {
+public:
+  ElaborationPass() : ElaborationPassBase() {}
+  ElaborationPass(const ElaborationPass &other) : ElaborationPassBase(other) {}
+  ElaborationPass(const rtg::ElaborationOptions &options)
+      : ElaborationPassBase(), options(options) {
+    if (options.seed.has_value())
+      seed = options.seed.value();
+    debugMode = options.debugMode;
+  }
+  void runOnOperation() override;
+  void cloneTargetsIntoTests();
+
+private:
+  ElaborationOptions options;
+  Pass::Option<unsigned> seed{*this, "seed",
+                              llvm::cl::desc("Seed for the RNG.")};
+  Pass::Option<bool> debugMode{*this, "debug",
+                               llvm::cl::desc("Enable debug mode."),
+                               llvm::cl::init(false)};
```

Comment:
Not sure what the best thing is, but I don't think it makes sense to store this duplicate information. Maybe just copy it out of the ElaborationOptions and in to the Pass::Options, which seems to be what happens for ODS declared options.